### PR TITLE
Integration of xen-troops patches

### PR DESCRIPTION
**Pull request by svlad-90**
Integration of the xen-troops patches on top of the Android U-Boot allows running Android using Virito devices within the Xen hypervisor domain.

### CODE REVIEW

File: cmd/bcb.c
Comment (lorc) [inline cmd/bcb.c]: Looks like some rebase artifact. Could you please remove unnecessary whitespace changes?

I can't see them in the original commit.

File: include/configs/xenguest_arm64.h
Comment (lorc) [inline include/configs/xenguest_arm64.h]: Missing EOL at EOF. Looks like this is unnecessary change.

File: arch/arm/dts/xenguest-android-arm64.dts
Comment (lorc) [inline arch/arm/dts/xenguest-android-arm64.dts]: EPAM 2023?

File: xen-guest-android-virtio.fragment
Comment (lorc) [inline xen-guest-android-virtio.fragment]: Missing EOL at EOF. There should be some setting in your editor to automatically add these.

File: arch/arm/cpu/armv8/start.S
Comment (lorc) [inline arch/arm/cpu/armv8/start.S]: Extra whitespace changes all around this patch. It makes review process difficult. 

Please refrain from changes that are not related to intended patch purpose.

File: arch/arm/cpu/armv8/start.S
Comment (lorc) [inline arch/arm/cpu/armv8/start.S:27]: `_start:` missing? I checked couple of "boot0.h" files, none of them define `_start` symbol

**Review by lorc — COMMENTED** [review]
The purpose of rebasing patches is not to resolve merge commits, but to produce valid codebase.  Looking at this PR, I can say that it wasn't accomplished: you are adding a bunch of patches that are fixing original rebased patches.  This is incorrect and provides a broken git history.  For example, if I checkout at "Add xen-specific targets to the BAZEL build" and try build, it fails with an error. 

You need to review your changes and squash appropriate fixes into original commits by Andrii to provide complete working history of changes.

Please remember, that no patch should break the build.

File: cmd/bcb.c
Comment (svlad-90) [reply cmd/bcb.c]: Hi @lorc, in this case, in the original commit author has mixed spaces and tabs in the source code. It was wrong ( even if it has passed a review ). Tabs are usually used in U-Boot. I've fixed that.

File: include/configs/xenguest_arm64.h
Comment (svlad-90) [reply include/configs/xenguest_arm64.h]: Hi @lorc, done.

File: arch/arm/dts/xenguest-android-arm64.dts
Comment (svlad-90) [reply arch/arm/dts/xenguest-android-arm64.dts]: Hi @lorc, done.

File: xen-guest-android-virtio.fragment
Comment (svlad-90) [reply xen-guest-android-virtio.fragment]: Hi @lorc, done.

Also, thanks for the hint. I've adjusted the settings in Sublime to add a new line on the "Save" operation.

File: arch/arm/cpu/armv8/start.S
Comment (svlad-90) [reply arch/arm/cpu/armv8/start.S]: Hi @lorc, I guess you mean "extra new lines", and not the extra "white spaces".

I've removed them. Done.

But, in general, from my point of view, it makes the code more readable in this case. I still believe in this. But you are the maintainer, so I obey. ))

File: arch/arm/cpu/armv8/start.S
Comment (svlad-90) [reply arch/arm/cpu/armv8/start.S:27]: Hi @lorc, I've checked them all.

|File link|Whether it has _start defined?|
|---|---|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-imx8/boot0.h|No|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-px30/boot0.h|Only for non ARM64, through include of [arch-rockchip/boot0.h](https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h#L53)|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-sunxi/boot0.h|No|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rk3036/boot0.h|Only for non ARM64, through include of [arch-rockchip/boot0.h](https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h#L53)|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rk3128/boot0.h|Only for non ARM64, through include of [arch-rockchip/boot0.h](https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h#L53)|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rk3188/boot0.h|Only for non ARM64, through include of [arch-rockchip/boot0.h](https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h#L53)|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rk322x/boot0.h|Only for non ARM64, through include of [arch-rockchip/boot0.h](https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h#L53)|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rk3288/boot0.h|Only for non ARM64, through include of [arch-rockchip/boot0.h](https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h#L53)|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rk3308/boot0.h|Only for non ARM64, through include of [arch-rockchip/boot0.h](https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h#L53)|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rk3328/boot0.h|Only for non ARM64, through include of [arch-rockchip/boot0.h](https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h#L53)|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rk3368/boot0.h|Only for non ARM64, through include of [arch-rockchip/boot0.h](https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h#L53)|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rk3399/boot0.h|Only for non ARM64, through include of [arch-rockchip/boot0.h](https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h#L53)|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rv1108/boot0.h|Only for non ARM64, through include of [arch-rockchip/boot0.h](https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h#L53)|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-bcm235xx/boot0.h|Yes|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-bcm281xx/boot0.h|Yes|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/include/asm/arch-rockchip/boot0.h|Yes|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/mach-rmobile/include/mach/boot0.h|Yes|
|https://github.com/xen-troops/u-boot/blob/master/arch/arm/mach-socfpga/include/mach/boot0.h|Yes|

For the majority of cases, there would be no "_start:" defined for ARM64. So, it seems, that we should add "_start:" here. I'll do this.

**Comment by svlad-90** [comment]
> The purpose of rebasing patches is not to resolve merge commits, but to produce valid codebase. Looking at this PR, I can say that it wasn't accomplished: you are adding a bunch of patches that are fixing original rebased patches. This is incorrect and provides a broken git history. For example, if I checkout at "Add xen-specific targets to the BAZEL build" and try build, it fails with an error.
> 
> You need to review your changes and squash appropriate fixes into original commits by Andrii to provide complete working history of changes.
> 
> Please remember, that no patch should break the build.

Hi @lorc. I think I've got your point. I'll try to follow the provided instructions. I guess, it will take me several hours to finish this activity. I'll let you know once it is finished with an additional comment.

File: cmd/bcb.c
Comment (svlad-90) [reply cmd/bcb.c]: It was agreed, that we should separate formatting change from functional changes. Thus, wither move it to a separate commit after this one, or remove this change at all.

Comment (svlad-90) [reply cmd/bcb.c]: Hi @lorc ,

Replacement of spaces with tabs was removed.

Done.

File: arch/arm/cpu/armv8/start.S
Comment (svlad-90) [reply arch/arm/cpu/armv8/start.S:27]: Hi @lorc, done.

**Comment by svlad-90** [comment]
Hi @lorc, all changes are done.

PR now consists of 8 commits instead of ~13 at the beginning. Most of the changes from the 5 disappeared commits were squashed into Andrii's commits.

Each separate commit was checked to be built successfully with:

- xenguest_android_arm64_defconfig
- xenguest_arm64_defconfig
- Bazel build with 'xen-guest-android-virtio_aarch64_dist' target ( last 3 commits, where this is applicable )

It is confirmed that Android is still up and running after this PR.

Please, have a look at the updated PR.

File: configs/xenguest_arm64_defconfig
Comment (lorc) [inline configs/xenguest_arm64_defconfig]: Thanks. I have (probably) no objections regarding commit diff, but I have comment about the commit message. It is unclear why you are doing those changes.  I want you to read section "Describe your changes": 

https://www.kernel.org/doc/html/latest/process/submitting-patches.html#describe-your-changes

Yes, this is the Linux manual, but it describes requirements that are used in virtually all embedded open source projects. You can ignore the Linux-related parts, but please stick to all other rules. 

For this particular commit message, you also need to drop "pvblock" word, as it is misleading: there is nothing pvblock-specific in this commit. Also, it is not clear what exactly you are trying to accomplish, so it is hard to review this change properly.

File: common/avb_pubkey.c
Comment (lorc) [inline common/avb_pubkey.c]: Whoa!

Should it be auto generated based on keys used? It is quite strange to include some public key, in the "common" section without providing instructions how to obtain private part of it. 

In my understanding this should be something deeply product-specific, but I can be wrong there.

Comment (lorc) [reply common/avb_pubkey.c]: Actually, I can see that "common/Makefile" issues `bin2c` tool to generate this file. Why it is included in this commit?

File: arch/arm/cpu/armv8/start.S
Comment (lorc) [reply arch/arm/cpu/armv8/start.S:27]: Thanks. But why this change is in "Add xen-specific targets to the BAZEL build" commit?

For me it looks like something that should be in a separate commit, probably worth upstreaming.

**Review by lorc — CHANGES_REQUESTED** [review]
Thanks for the changes. It is in a much better shape now. But there are still couple of relatively minor issues to fix.

File: arch/arm/cpu/armv8/start.S
Comment (svlad-90) [reply arch/arm/cpu/armv8/start.S:27]: Hi @lorc, it is in the same commit, because the Bazel build is failing without this change. As we agreed - each commit should be compilable. That motivated me to place it in such a way. The thing I can do is place the commit with the compilation issue fix before the "Add xen-specific targets to the BAZEL build". That will ensure atomic, still buildable commits. I will do that tomorrow.

Comment (lorc) [reply arch/arm/cpu/armv8/start.S:27]: Yep, this is the right thing to do.

Comment (svlad-90) [reply arch/arm/cpu/armv8/start.S:27]: Hi @lorc, done.

File: common/avb_pubkey.c
Comment (svlad-90) [reply common/avb_pubkey.c]: Hi @lorc, the file was added accidentally. I've removed it.

Done.

File: configs/xenguest_arm64_defconfig
Comment (svlad-90) [reply configs/xenguest_arm64_defconfig]: Hi @lorc,

The original commit which introduced almost all those flags is this one:
https://github.com/xen-troops/u-boot/commit/91a77926c6f3ced01c2c2bcf926dd4c8e10136cd
It was upstreamed.

But this file has a big history in upstream, which excludes the majority of the flags commit after commit:
https://github.com/xen-troops/u-boot/commits/xenvm-trout-main/include/configs/xenguest_arm64.h

I'll check which of the flags are still relevant and will remove all the others.

Comment (svlad-90) [reply configs/xenguest_arm64_defconfig]: Hi @lorc,

After the review of what was upstreamed, I've figured out, that I need no any of those flags at all. This patch of Andrii was upstreamed, and then further patches in the upstream moved almost everything to:

- KConfig default values
- xenguest_arm64_defconfig ( yes, some of my added flags were already there )

Done.

P.S. I will review the Android-specific flags from the same perspective. Those flags which would be automatically added through KConfig are not needed to be presented in the defconfig.

Comment (svlad-90) [reply configs/xenguest_arm64_defconfig]: The same flags for xenguest_android_arm64_defconfig were reviewed and changed.

**Comment by svlad-90** [comment]
Hi @lorc, all changes were done. Please have another look at it.

**Comment by svlad-90** [comment]
Hi @lorc, please, do not merge yet. Let me check that system is up and running, after all the today's changes.

**Review by lorc — APPROVED** [review]
Thanks. It is in a much better shape now.

One last comment: Patch "Add xen-specific targets to the BAZEL build" has outdated description, as it mentions "- Fixed comilation issue with redefinition of the '_start' symbol
( clang specific )". Please fix the description and add my r-b tag to all commits:

`Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

**Comment by svlad-90** [comment]
Hi @lorc, the changes were tested on the target. U-Boot still works fine for booting Android on our virtio set-up.

Regarding r-b and changing the specific commit message - ok. I'll do that and will let you know.

**Comment by svlad-90** [comment]
Hi @lorc, the changes are done.

Thanks for the review! ))

@arminn, @lorc, please, merge the change.

**Comment by lorc** [comment]
@arminn, the honor is yours :)

**Pull request merged**
